MSBuild Versioning Task

JohnTortugo opened the pull request:
Continuation from this https://github.com/dotnet/arcade/pull/137 - that PR got too big and the page was frequently failling to load;

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Util/Microsoft.DotNet.Build.Tasks.Util.csproj:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="RoslynTools.RepoToolset">
```
I'm not sure this is worth its own project. Maybe if we get enough stuff we can do it but for now I would suggest just having the one common source file included in the other project.

JohnTortugo replied:
OK. Removed.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+{
+    public class GenerateVersion : BuildTask
+    {
+        public static string SHA = String.Empty;
```
I'd make these static properties private so that that are only used by this task library.

weshaggard replied:
I would name these s_shortSha, s_shortDate, s_revision to follow coding guidelines https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

weshaggard replied:
They are intentionally static to cache the git commands across all the different projects trying to access them.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+
+            GeneratedShortDate = AdjustPadding(Date);
+            GeneratedShortSha = SHA;
+            GeneratedRevision = AdjustPadding(Revision);
```
We want a different padding for Revision, we don't want it to be 5. Maybe we just fix the padding to 5 and 2 and just have the one IncludePadding property for now and if we ever need to change the lengths we can expose a property for each at that time.

JohnTortugo replied:
OK. Will do it.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GitInfo.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Microsoft.DotNet.Build.Tasks.Versioning
+{
+    public class GitInfo : ToolTask
```
Why does this need to be a Task? Also I would make it private unless we make it more generalized for public consumption.

JohnTortugo replied:
I don't need to be a Task. I understood that this would be just a temporary implementation while the #136 doesn't get finished. Should we really go for something else since this will be deprecated soon?

weshaggard replied:
I don't mind it being a task temporarily but please make it private so that we don't end up with any dependencies on it from outside.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GitInfo.cs:
```diff
@@ -0,0 +1,44 @@
+
+        protected override string GenerateCommandLineCommands()
+        {
+            return "log -1 --format=\"%h %ci\"";
```
Why did you choose `ci` for the date? Will that correctly get us a consistent date across time zones and that can be parsed by DateTime.TryParse?

JohnTortugo replied:
It's the committer date in ISO format. I think committer date is best than author date, but it'll probably be the same in our setups. It includes the timezone info and is parsed by DateTime.TryParse.  Some documentation about it: https://git-scm.com/docs/git-show other people also use the same: https://github.com/loresoft/msbuildtasks/blob/master/Source/MSBuild.Community.Tasks/Git/GitCommitDate.cs

rainersigwald replied:
I agree that committer date is preferable to author date. They will diverge in rebase-heavy workflows (like we use on our team).

Looking at that doc though, should it be `%cI`? 

> %ci: committer date, ISO 8601-like format
> %cI: committer date, strict ISO 8601 format

JohnTortugo replied:
From what I can find there are only minor differences; But to be on the safe side I'll make it strict.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,8 @@
+<Project ToolsVersVersioningn="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
```
Where is the targets file?

JohnTortugo replied:
@weshaggard Do you mean the file that has the <UsingTask..> declarations? If yes, then this is the file and I'll rename it accordingly.

JohnTortugo replied:
@rainersigwald AFAIU this package will be included by projects whenever they want to generate a version string. This package will generate pieces of the version string and the calling package/project is responsible for assembling the string. The version string format is defined by this PR https://github.com/dotnet/arcade/pull/108.

weshaggard replied:
I was expecting to see the targets/props file that does all the work of calling this target, similar to your earlier PR.

rainersigwald replied:
What I think would be useful here is a small example. Suppose I have a trivial project, 

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
</Project>
```

and I want to use this. What does the project look like after having done so?

JohnTortugo replied:
@weshaggard Ok, got it.

JohnTortugo replied:
@rainersigwald , that is a great question. Thanks. I am not quite sure how these packages will be delivered to "Arcade on-boarded" projects. I think it's as Nuget packages. So I think at the end it will be something like this:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
  </PropertyGroup>
  
  <ItemGroup>
    <PackageReference Include="Microsoft.DotNet.Build.Tasks.Versioning" Version="1.0.0" />
  </ItemGroup>
  
  <Import Project="$(Somewhere)\Microsoft.DotNet.Build.Tasks.Versioning.props" />
  <Import Project="$(Somewhere)\Microsoft.DotNet.Build.Tasks.Versioning.targets" />
  
  <Target Name="Build" DependsOnTargets="GenerateVersion">
  		<Message Text="Starting build: $(ShortDate).$(Revision)-$(ShortSha)" />
    </Target>
</Project>
```
@mmitche , @weshaggard , I am also puzzled by this. Could you clarify this point?

alexperovich replied:
As I am now working on the initial "sdk" for arcade, the way this package will be consumed is by nothing but the following
```xml
  <ItemGroup>
    <PackageReference Include="Microsoft.DotNet.Build.Tasks.Versioning"/>
  </ItemGroup>
```

Either in the arcade sdk package as a default reference or inside the project file.

JohnTortugo replied:
In that case, how do I know the path to the .props and .target files?

rainersigwald replied:
You don't have to! NuGet handles it for you, as long as they're named `build/{PackageName}.props` and `.targets`. Docs: https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package#including-msbuild-props-and-targets-in-a-package

JohnTortugo replied:
Thanks! I'll take a look

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            {
+                GitInfo gitInfo = new GitInfo();
+                gitInfo.BuildEngine = this.BuildEngine;
+                gitInfo.Execute();
```
What happens if git cannot be found? It seems like we would simply throw an exception here.

JohnTortugo replied:
I am adding code to handle the exception and print a message in case of problems. I think better handling of this should be done in the future Git package itself.

weshaggard replied:
I think this still needs some handling here, we should at least catch the exception and correctly default the properties on gitinfo.

JohnTortugo replied:
I added the code to catch the exception, had not pushed it yet.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+                gitInfo.BuildEngine = this.BuildEngine;
+                gitInfo.Execute();
+
+                var BuildIdParsedCorrectly = GetDateAndRevisionFromBuildId(OfficialBuildId, out var TempDate, out var TempRevision);
```
TempDate => tempDate, TempRevision => tempRevision

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            return true;
+        }
+
+        private bool GetDateAndRevisionFromBuildId(string buildId, out string Date, out string Revision)
```
make Date and Revision lowercase.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            Date = String.Empty;
+            Revision = String.Empty;
+
+            if (String.IsNullOrEmpty(buildId)) return false;
```
Do this check before you run the regex.Match.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+
+            if (String.IsNullOrEmpty(buildId)) return false;
+
+            if (match.Success && match.Groups.Count > 2)
```
Not sure we need the match.Groups.Count check.

JohnTortugo replied:
In this case we don't need it.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            return false;
+        }
+
+        private string CreateShortDate(DateTime BuildDate)
```
lowercase input parameter.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            if (months > 0)
+            {
+                return IncludePadding
+                    ? string.Format("{0}{1}", months.ToString("D" + (Padding - 2)), BuildDate.Day.ToString("D2"))
```
looks like you are going to do the padding twice, once here and then once in your AdjustPadding function later.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+
+        private string AdjustPadding(string input)
+        {
+            if (IncludePadding)
```
To normalize you need to always trim and then add the padding to the desired length after that.

weshaggard commented:
@JohnTortugo what is the status of this PR?

JohnTortugo commented:
After two weeks on FR I went on vacation. I'll be back to work next monday and then I'll resume this.

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Microsoft.DotNet.Build.Tasks.Versioning.csproj:
```diff
@@ -0,0 +1,39 @@
+    <TargetFrameworks>netcoreapp2.0;$(NetFxTfm)</TargetFrameworks>
+    
+    <IsPackable>true</IsPackable>
+    <NuspecFile>$(MSBuildProjectName).nuspec</NuspecFile>
```
I was trying to follow Microsoft.DotnetBuild.Tasks.IO as an example. I removed this .nuspec file.

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Microsoft.DotNet.Build.Tasks.Versioning.csproj:
```diff
@@ -0,0 +1,39 @@
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == '$(NetFxTfm)'">
+    <Reference Include="Microsoft.Build.Utilities.v4.0" />
```
I chatted on Skype with @rainersigwald and a question popped up: @mmitche / @weshaggard do we need to target NetFxTfm (4.5) ?

weshaggard replied:
Yes we will need to target a version of .NET Framework. I think right now we are targeting net461. The plan for our task libraries will be to target a version of .NET Core and a version of .NET Framework.

JohnTortugo replied:
OK. Thanks. As per Directory.Build.props NetFxTfm is net45.

JohnTortugo replied:
Talked with @rainersigwald and he suggested me to use MSBuild packages version 14.3.0 for Net Framework 4.5. So, for now we're going to use 14.3.0 for Net Framework and 15.7.179 for Net Core.

Once we bump TFM to 4.6 we can refactor this file and reference 15.7.179 for Net Core and Net Framework.

rainersigwald replied:
You should be able to use 14.3.0 for both now, and update to something more modern later.

JohnTortugo replied:
Wouldn't it be better to use the latest where we can?

rainersigwald replied:
You can, but there shouldn't be any meaningful difference in the API surface for this task, so no advantage to the higher version.

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Util/BuildTask.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public abstract partial class BuildTask : ITask
```
Changed to use Task.

JohnTortugo commented:
Pushed some changes after your feedback. PTAL:

- Solved merge conflicts with master
- Renamed several fields, parameters, etc.
- Removed the Tasks.Util project
- Added exception handling around GitInfo task call
- Fixed the padding for date and revision
- Made GitInfo class internal to assembly
- Added .targets file
- Fixed bug at checking for null during parsing of BuildInfo
- Fixed use of Padding function
- Normalized trimming in padding function
- Changed GenerateVersion to use MSBuild Task as base class
- Deleted BuildTask class
- Removed nuspec file
- Moved .target/.props file to build/ folder
- Added some comments to initialization of some fields with default values
- Fixed reference to MSBuild packages

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,157 @@
+                    }
+                    else
+                    {
+                        s_shortDate = gitInfo.HeadCommitDate != default(DateTime) ? CreateShortDate(gitInfo.HeadCommitDate) : 
```
I think I would move the default(DateTime) check into CreateShortDate function and if it is set to default the set it to UtcNow.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,157 @@
+        /// <summary>
+        /// The size of the padding for the GeneratedShortDate and GeneratedRevision fields.
+        /// </summary>
+        private readonly int datePadding = 5;
```
Add "_" to these members so it is clear these are members of this class.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,157 @@
+
+            if (String.IsNullOrEmpty(buildId)) return false;
+
+            Regex regex = new Regex(@"(\d{8})[\-\.](\d+)$");
```
We don't need to do anything now but we might need to loosen the "8" at some point to allow for other formats.

JohnTortugo replied:
I'll add a comment there.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/targets/Microsoft.DotNet.Build.Tasks.Versioning.targets:
```diff
@@ -0,0 +1,11 @@
+<Project ToolsVersVersioningn="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Target Name="GenerateVersion">
```
Where do we hook this up? Also I assume you still plan to have a props file that defaults all the properties (as well as the doc) and concatenates the version strings? 

As a reference have a look at https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props and https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.targets they aren't exactly what you need but they are pretty close.

JohnTortugo replied:
"Where do we hook this up?" Do you mean how will this be called? I was thinking that it would be better if interested projects would register this target as a "DependsOnTargets", this way they can control exactly when it is called. Should I change it to run before other target - like GetAssemblyVersion, Build?

JohnTortugo replied:
> "concatenates the version strings" 

I was thinking that this package would only expose pieces of the version string and clients would be responsible for constructing the string itself. Isn't that so? May be we should talk about this then.

JohnTortugo replied:
> "I assume you still plan to have a props file that defaults all the properties"

I am pushing a .props file but it only has a include for the DLL and a UsingTask directive. As of now all defaults are set on the GenerateVersion.cs file; Maybe we need to change this if this task also has to construct the whole version string.

weshaggard replied:
>I was thinking that this package would only expose pieces of the version string and clients would be responsible for constructing the string itself. Isn't that so? May be we should talk about this then.

Ideally what we want is for folks to reference this package from an sdk style project and they all the necessary versioning hooked up automatically. We don't want repo's to need to do extra work to hook this up. 

If we decide to merge this task in to the root arcade sdk package then we could include the targets/props necessary in that package but if we remain separate then it should happen in the targets/props in this package.

weshaggard replied:
For reference the dotnet SDK uses the following properties:

VersionSuffix: https://github.com/dotnet/sdk/blob/f0ead5620a318ca57aa1a9e3e7b18db5d6984eaf/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.DefaultAssemblyInfo.targets#L20 which is used to compute the version.

That Version property is used as the nuget package version and then pulled into the FileVersion and InformationVersion in the assembly here https://github.com/dotnet/sdk/blob/f0ead5620a318ca57aa1a9e3e7b18db5d6984eaf/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L160. all those need to be updated based on the version we are computing here.

JohnTortugo commented on test/Microsoft.DotNet.TestsUtil/Microsoft.DotNet.TestsUtil.csproj:
```diff
@@ -0,0 +1,18 @@
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkVersion)" />
+    <PackageReference Include="xunit" Version="$(XunitVersion)" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="$(XunitVersion)" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="15.1.548" />
```
I'll do that. However, I'll defer working on this test project again until we get settled on the workings of the versioning task.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -62,24 +62,28 @@ public override bool Execute()
             {
-                GitInfo gitInfo = new GitInfo();
-                gitInfo.BuildEngine = this.BuildEngine;
+                var HeadCommitSHA = "NOSHA";
```
Coding guidelines: You shouldn't use var here, you should explicitly type the variable. Also the first char should be lowercase.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -62,24 +62,28 @@ public override bool Execute()
                 {
-                    Log.LogError($"Problems retrieving branch information from Git: {ex.Message}");
-                    return false;
+                    Log.LogError($"Problems retrieving branch information from Git: {ex.Message}. Defaults for SHA and Commit date will be used.");
```
This shouldn't be an error as it will cause the task to fail which we don't want. I don't even think it is worth a warning it should probably just be a normal log message.

JohnTortugo replied:
Will change it.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,175 @@
+
+        private string CreateShortDate(DateTime buildDate)
+        {
+            if (buildDate != default(DateTime))
```
should be `buildDate == default(DateTime)`

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,175 @@
+                    Log.LogError($"Problems retrieving branch information from Git: {ex.Message}. Defaults for SHA and Commit date will be used.");
+                }
+
+                var buildIdParsedCorrectly = GetDateAndRevisionFromBuildId(OfficialBuildId, out var tempDate, out var tempRevision);
```
Please explicitly list the types for the variables and output parameters instead of using var.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,175 @@
+            }
+            else
+            {
+                Log.LogError("Error: Comparison date is in the same month as the seed date");
```
I think the error should be "Error: Baseline date is greater then the build date".

JohnTortugo commented:
Anything more comments?

JohnTortugo commented:
@dotnet-bot test this please

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,25 @@
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <OfficialBuildId>$(BUILD_BUILDNUMBER)</OfficialBuildId>
```
I'm not sure we want to set OfficialBuildId in this props file, it is generally a property that is passed in.

JohnTortugo replied:
Removed.

weshaggard commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.targets:
```diff
@@ -0,0 +1,31 @@
+      <!-- IncludePadding also means that we are using SemVer1.0 format. -->
+      <VersionSuffix Condition="'$(IncludePadding)' == 'true'">$(VersionSuffix.Replace('+', '-').Replace('.', '-'))</VersionSuffix>
+      
+      <Version>$(VersionPrefix)</Version>
```
We are going to need to set more versions. At least "FileVersion" and "PackageVersion".

JohnTortugo replied:
Will FileVersion have only the the ShortDate as suffix? This seemed to be the case here: https://github.com/dotnet/roslyn-tools/blob/bfdd2ce65675faf22f5869dfee85e5350da7bfed/src/RepoToolset/tools/Version.targets#L20

JohnTortugo replied:
Could you please confirm the format for PackageVersion? I didn't find it's definition on roslyn-tools or dotnet-sdk; found something on buildtools but I am not sure of the format: https://github.com/dotnet/buildtools/search?l=XML&p=1&q=packageversion

weshaggard replied:
By default PacakgeVersion is set to Version https://github.com/NuGet/NuGet.Client/blob/50af5271b98ac5cb2896a707569bc4cd1e87a017/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L28 but that is done before we set Version here. So we just need to set PackageVersion to Version here. I would have expected your testing to catch this as the package wouldn't have had the correct version.

As for FileVersion it needs to be in a format similar to https://github.com/dotnet/buildtools/blob/5254ae0c85b37d36695ae3ff4948f626b21d3bb9/src/Microsoft.DotNet.Build.Tasks/PackageFiles/versioning.targets#L24. We should also have a look to see if we need to set AssemblyFileVersion I'm not sure without some verification of the logs. 

Please make sure your testing has the correct assembly, file, and package version after you build with this package.

tmat replied:
PackageVersion *is* the version we are constructing here.

---
In reply to: [198263751](https://github.com/dotnet/arcade/pull/178#discussion_r198263751) [](ancestors = 198263751)

tmat replied:
It looks like dotnet SDK expects `Version` to be calculated at evaluation time.

We can overwrite it by a target, but we are risking that some other targets already read the old value. We'd need to keep track of these targets and chain them into `BeforeTargets` of this `GenerateVersionString` target.

E.g. `GenerateNuspec` target reads `PackageVersion` which is initialized from `Version` during evaluation phase. If we overwrite `Version` we need to overwrite `PackageVersion` as well and add `GenerateNuspec` to `BeforeTargets`.

It would imo be better if we could calculate VersionSuffix in evaluation time.  

We can get the full commit SHA from `$(BUILD_SOURCEVERSION)` on the server. Can we just take substring [0, 10) from it? 

See [https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L42](https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L42) for evaluation-time version calculation (might need some adjustments).

Re 
> Local dev builds - Date of current git HEAD
in the versioning spec. Why do we need this in dev builds?  
Why not just skip this version part in dev builds?


---
In reply to: [198269711](https://github.com/dotnet/arcade/pull/178#discussion_r198269711) [](ancestors = 198269711)

tmat replied:
Re FileVersion - the ShortDate won't work as the last component of FileVersion, since it will almost certainly be > 655. We do adjust short date in RepoToolset to account for that [https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L42](https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L42) 

---
In reply to: [198311990](https://github.com/dotnet/arcade/pull/178#discussion_r198311990) [](ancestors = 198311990,198269711)

weshaggard replied:
>Re FileVersion - the ShortDate won't work as the last component of FileVersion, since it will almost certainly be > 655. 

The value needs to be able to fit into a ushort so < 65535. We also need to make sure this value is computed by an offset to ensure continuous versioning from already shipped products. 

>It looks like dotnet SDK expects Version to be calculated at evaluation time.

I agree that this is potentially an issue but I'm not convinced we should be doing all the inline computation as that ends up being costly for lots of projects without any caching and I think it is harder to maintain to handle all the different cases we want. Coding inline in msbuild is definitely not pretty or something I generally recommend.

`$(BUILD_SOURCEVERSION)` that will work in some cases but not all. There are cases where we have wrapper build definitions where this property isn't set to the repo we are actually building. Also that wouldn't work for dev builds, assuming we decide to keep it for those.

tmat replied:
> I agree that this is potentially an issue but I'm not convinced we should be doing all the inline computation as that ends up being costly for lots of projects without any caching 

We have optimized the calculations to avoid conversion exceptions etc. I agree it's not ideal, but I think it's better than potentially introduce inconsistency into the build. It would be nice if msbuild supported custom functions that are executable in evaluation time. 

@rainersigwald @nguerrera Any opinions? To recap context, it seems like the .NET SDK is expecting `VersionSuffix` (and `Version`) to be calculated at evaluation time. It's non-trivial/slow to do the calculation using existing built-in methods. See https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L38. Should the SDK support calculating Version by a target? Should msbuild support custom functions callable at evaluation time?

tmat replied:
@weshaggard 

> $(BUILD_SOURCEVERSION) that will work in some cases but not all. There are cases where we have wrapper build definitions where this property isn't set to the repo we are actually building.

What are the cases? Source build? Why can't the orchestrated build set the variable properly?

> Also that wouldn't work for dev builds, assuming we decide to keep it for those.

For dev builds I think it's sufficient to set `BUILD_SOURCEVERSION` to some fixed value (conditionally, if not set yet). That would exercise all the build logic just like it would happen on CI machine.

tmat replied:
More thoughts re dev build and commit sha -- I think it might actually be confusing to use the real commit sha during dev build. The reason is that it the SHA doesn't reflect the state of the local repo if there are any files changed. On CI it does reflect the exact state of the repo since there are no modified files.

weshaggard replied:
>What are the cases? Source build? Why can't the orchestrated build set the variable properly?

In the past there have been cases were we had to wrap calls into clone and redirect to a different repo then what the build definition is pointing at, in those cases build.sourceversion isn't set to what we need. The cases that I recall are things were we used our old PipeBuild system and in cases were we needed to build potentially from different repo's (i.e. github and vsts) from the same build definition. I'm not sure how many of those will still be around after we convert but those are examples were I know build.sourceversion wouldn't be set correctly today.

>More thoughts re dev build and commit sha

I could get behind not including the sha for dev builds, @mmitche if you agree perhaps we can make an update the the versioning spec to call those outs.

mmitche replied:
Yeah I'm fine with not including the sha for dev builds.

The only case I know of where we have built another repo from the initial clone is pipebuild, and we intend to get rid of that.  VSTS itself doesn't allow for building of repos other than what the definition was configured with (can't change at queue time).  They may change up their features to allow for multiple repos to be cloned in the future, which would change the environment variable situation.

weshaggard commented:
@tmat would you mind having a look over this PR as well?

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Microsoft.DotNet.Build.Tasks.Versioning.csproj:
```diff
@@ -0,0 +1,40 @@
+
+    <!-- Package Properties -->
+    <Title>Packaging Targets and Tasks used to create versioning information</Title>
+    <MinClientVersion>2.5.0</MinClientVersion>
```
><MinClientVersion>2.5.0</MinClientVersion> [](start = 4, length = 42)

Why is this needed? I haven't seen any of our other packages to specify this value.

JohnTortugo replied:
I remember that I included this after I got a warning message. I removed it now and I didn't get the message. So I am leaving it out until I see need for it again.

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/Microsoft.DotNet.Build.Tasks.Versioning.csproj:
```diff
@@ -0,0 +1,40 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+		<Content Include="$(MSBuildThisFileDirectory)build\*.targets">
```
>	 [](start = 1, length = 1)

Use spaces, not tabs

tmat replied:
>$(MSBuildThisFileDirectory) [](start = 20, length = 27)

`$(MSBuildThisFileDirectory)` is not necessary.

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,25 @@
+
+    <ShortDate></ShortDate>
+    <Revision></Revision>
+    <ShortSha></ShortSha>
```
Why set empty properties?

JohnTortugo replied:
Reminiscent from old version of this PR... thanks.

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,25 @@
+      We use Semantic Versioning 2.0 which doesn't require padding on identifiers. 
+      However, it's possible to rever to SemVer 1.0 and padding is required.
+    -->
+    <IncludePadding>false</IncludePadding>
```
>IncludePadding [](start = 5, length = 14)

This property name is too general. How about renaming to `<SemanticVersioningV1>` ?

JohnTortugo replied:
Also, do we really need BaselineDate as a parameter? Will we need to use a custom comparison date for calculating the short date field?

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,25 @@
+    <OfficialBuildId>$(BUILD_BUILDNUMBER)</OfficialBuildId>
+
+    <!-- The default prefix assumes Major=1, Minor=0 and Patch=0 -->
+    <VersionPrefix>1.0.0</VersionPrefix>
```
>VersionPrefix [](start = 5, length = 13)

This is already default set by .NET SDK

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.targets:
```diff
@@ -0,0 +1,31 @@
+
+  <Target Name="GenerateVersionString" BeforeTargets="GetAssemblyVersion">
+    <GenerateVersion OfficialBuildId="$(OfficialBuildId)" BaselineDate="$(BaselineDate)" IncludePadding="$(IncludePadding)">
+      <Output TaskParameter="GeneratedShortDate" PropertyName="ShortDate" />
```
>ShortDate [](start = 63, length = 9)

Prefix these property names with `_`, which denotes "private" property.

tmat commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersionComponents.cs:
```diff
@@ -0,0 +1,175 @@
+
+            if (emptySha || emptyDate || emptyRevision)
+            {
+                var HeadCommitSHA = "NOSHA";
```
We can validate everything locally -- set the variables to some fixed versions:
```
BUILD_BUILDNUMBER=20180102.03
BUILD_SOURCEVERSION=12345678901234567890123456789012
```

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.targets:
```diff
@@ -0,0 +1,36 @@
+      <PreReleaseLabel Condition="'$(PreReleaseLabel)' == ''">dev</PreReleaseLabel>
+      
+      <!-- This handles the Dev/Daily case of the versioning plan. -->
+      <VersionSuffix Condition="'$(PB_IsStable)' != 'true'">$(PreReleaseLabel).$(_ShortDate).$(_Revision)+$(_ShortSha)</VersionSuffix>
```
So, should I go ahead and remove the "+$(_ShortSha)" part? @weshaggard @tmat

tmat replied:
The point of not including SHA in dev build was to be able to use BUILD_SOURCEVERSION environment variable set by VSTS to get the commit SHA and thus avoid having to call a task to construct the version number.

JohnTortugo replied:
What about the other fields, like ShortDate, that are computed in relation to the repo last commit date? Should we assume there will always be an OfficialBuildId to extract it from?

tmat replied:
Yes, I'd think so. That's what we do in Roslyn and RepoToolset.

I'd prefer calculating the short date like so unless there is some issue with this approach that we need to address:
https://github.com/dotnet/roslyn-tools/blob/master/src/RepoToolset/tools/Version.props#L25-L42

weshaggard replied:
We could remove shortsha from dev built we still need it for daily so we would need another way to detect that.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/build/Microsoft.DotNet.Build.Tasks.Versioning.props:
```diff
@@ -0,0 +1,15 @@
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <_MicrosoftDotNetBuildTasksVersioningTaskAssembly Condition=" '$(MSBuildRuntimeType)' == 'Core' ">$(MSBuildThisFileDirectory)..\lib\netcoreapp2.0\Microsoft.DotNet.Build.Tasks.Versioning.dll</_MicrosoftDotNetBuildTasksVersioningTaskAssembly>
+    <_MicrosoftDotNetBuildTasksVersioningTaskAssembly Condition=" '$(MSBuildRuntimeType)' != 'Core' ">$(MSBuildThisFileDirectory)..\lib\$(NetFxTfm)\Microsoft.DotNet.Build.Tasks.Versioning.dll</_MicrosoftDotNetBuildTasksVersioningTaskAssembly>
```
I don't think you can ship `$(NetFxTfm)` in the package--it's not defined normally. I'd probably just hardcode it.

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersionComponents.cs:
```diff
@@ -0,0 +1,175 @@
+    {
+        /// <summary>
+        /// We use these fields as a form of caching for the generated
+        /// output fields. That's why they are static properties.
```
This isn't the best way to do this. If you build, change commits, and build again with the same (reused) MSBuild nodes, you'll get a wrong answer.

You can use [RegisterTaskObject](https://github.com/Microsoft/msbuild/blob/f147a76a60bf9e516800f65f9614c914df0a4f15/src/Framework/IBuildEngine4.cs#L34-L62) with `RegisteredTaskObjectLifetime.Build` to get caching but expire it at the correct time, or you could drop caching entirely at the task level by building a single project from all projects that consume this package and returning results from that.

JohnTortugo replied:
I don't know if the scenario you describe (build-change-build) will happen, but seems likely. Can you comment @weshaggard ? If yes, I'll sync with @rainersigwald to do the required changes.

tmat replied:
@JohnTortugo It will definitely happen. 

Let's first resolve the question of whether we actually need to retrieve information from git or whether we can use VSTS variables (`BUILD_BUILDNUMBER` and `BUILD_SOURCEVERSION`) to construct the version string. If we use the environment variables then the question of optimizing this is moot.

rainersigwald replied:
It's definitely less of an issue for official build scenarios, but it's highly likely on developer desktops (and builds inside VS).

Personally, I don't like depending on the environment because I want reproducibility on developer desktops. If that's not a goal, it would definitely simplify things.

tmat replied:
I agree that the environment dependencies should be minimal. The proposal is that if `BUILD_BUILDNUMBER` and `BUILD_SOURCEVERSION` aren't defined we use some hard-coded values. So the entire build will be reproducible in developer builds. it will just produce packages with versions like `1.2.3-beta1-99999` or something like that. 
In projects that deploy VSIX packages during build we even force the assembly version to `42.42.42.42`, so that it's distinct from and higher than any official build version. This is to force the VSIX installer to replace the currently installed official version of the VSIX with the one built by the dev.

JohnTortugo replied:
@tmat sounds good

rainersigwald commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersionComponents.cs:
```diff
@@ -0,0 +1,175 @@
+                {
+                    GitInfo gitInfo = new GitInfo();
+                    gitInfo.BuildEngine = this.BuildEngine;
+                    gitInfo.Execute();
```
I don't like this manually-calling-a-task-inside-a-task pattern. I don't _know_ of anything that goes wrong when it's done, but I wouldn't be surprised.

JohnTortugo commented on src/Microsoft.DotNet.Build.Tasks.Versioning/src/GenerateVersion.cs:
```diff
@@ -0,0 +1,147 @@
+            GeneratedShortSha = SHA;
+            GeneratedRevision = AdjustPadding(Revision);
+
+            return true;
```
Thanks!

JohnTortugo commented on test/Microsoft.DotNet.Build.Tasks.IO.Tests/Microsoft.DotNet.Build.Tasks.IO.Tests.csproj:
```diff
@@ -1,5 +1,6 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
I don't.. and I don't know if we have a guide line for that. :-/

JohnTortugo commented:
Sorry to be harsh on this: Can you list specific things that are blocking (not an optimization) to have a first version of this PR merged?

JohnTortugo commented:
Update: a discussion about further changes on this PR is happening on an e-mail thread. Ping me if you want to be included..

JohnTortugo commented:
Closing this as it diverges substantially from the way we are going to implement the versioning.
